Update code PairDevice calls to work with network params.

cecille opened the pull request:
#### Problem
The setup code PairDevice functions had no way to pass in commissioning parameters, which meant it wasn't possible to commission BLE devices by code.

#### Change overview
- Adds a function to the auto commissioner to allow adding commissioning parameters incrementally
- Add commissioning parameters as a function parameter to the pair device calls for setup codes
- Adds pairing options to chip tool to allow this path

#### Testing
Chip tool commissioning M5 over BLE

sagar-apple commented:
This PR aligns all high-level commissioning pathways to use the auto commissioning state machine. In the case of a cli based human driven command this might make sense but it also forces the commissioner using the payload string APIs to gather network credentials for all possible networks prior to commissioning. 

It remains desirable for some use cases to only have PASE setup and to trigger commissioning at some point later. 

For example, what if you don't have Thread credentials available? What if you want to present an error when you realize that someone/something is trying to commission an accessory that your network cannot support? How do you even query what network technologies the accessory has without attempting to provision it with something. 

At least for now, we should allow the caller to specify auto commissioning when using the payload APIs and update SetupCodePairer to also just perform PASE if requested by the caller.

As an aside, pretty sure this breaks CHIPTool on iOS but that thing gets no love these days.

@msandstedt thoughts?

bzbarsky-apple reviewed (changes requested):
This PR forces getting network credentials before doing discovery and finding out whether they are even needed.

That's not acceptable, unfortunately.  We have to allow commissioners using SetUpCodePairer to do discovery and set up PASE and only then deal with getting network credentials as needed.

msandstedt commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -85,12 +85,14 @@ CommissioningParameters PairingCommand::GetCommissioningParameters()
 CHIP_ERROR PairingCommand::PairWithQRCode(NodeId remoteId)
 {
-    return CurrentCommissioner().PairDevice(remoteId, mOnboardingPayload);
+    CommissioningParameters commissioningParams = GetCommissioningParameters();
+    return CurrentCommissioner().PairDevice(remoteId, mOnboardingPayload, commissioningParams);
```
When we set some commissioning parameters and not others with the PairDevice call, what parameters are we leaving untouched?

If there are parameters we must always overwrite, and others we must always preserve, it feels like these should be in separate structures.

Overall, it seems like merging into the commissioning parameters like this is a good way to inadvertently leak state across commissioning sequences.

cecille replied:
That is a fair point. I actually sorted a way to do this without the Add command. I will update.

cecille commented:
I would be open to the idea that the setup code pairer triggers a pase connection only in all cases (BLE and on-network), and commission is called separately. But the accessor functions for that should be renamed to make that clear.

cecille commented:
Proposal:
1) I fix Michael's thing and remove the "add" command in the auto commissioner
2) PairDevice commands using the setup code pairer trigger commissioning using network credentials provided apriori. Once https://github.com/project-chip/connectedhomeip/pull/13829 goes in, you can just provide thread and wifi and the commissioner will select the appropriate one.
3) We add additional EstablishPASEConnection functions based off the setup code that trigger PASE only for folks that want that. That way we have all the combos.

sagar-apple commented:
>Was the intent to have set up code pairer do pase only? Because that's not how it was written originally. In the original instance, setup code pairer triggered commissioning and it is named as such.

You're completely right about this. The meaning of Pairing got separated from Commissioning and CASE at some point and now they're being aligned again so that explains the weirdness with naming all over the place.

It used to mean "PASE is ready and node is Commissioned" after which it was up to the commissioner to actually trigger the CASE (via get Connected Device) after providing network credentials. But that's old stuff now so let's ignore it. 

Your proposal for the most part sounds acceptable. 

>PairDevice commands using the setup code pairer trigger commissioning using network credentials provided apriori. Once Check feature map for network technology #13829 goes in, you can just provide thread and wifi and the commissioner will select the appropriate one.

I think the problem here is that it's a little bit backwards to handover everything before even starting to commission. 

>We add additional EstablishPASEConnection functions based off the setup code that trigger PASE only for folks that want that. That way we have all the combos.

This sounds good. Unfortunately, I'm not sure how to exercise this from the cli tooling. So that's something we'll have to think about.

msandstedt commented:
> I would be open to the idea that the setup code pairer triggers a pase connection only in all cases (BLE and on-network), and commission is called separately. But the accessor functions for that should be renamed to make that clear.

In my mind, the primary reason to include PASE with discovery (the SetUpCodePairer's current primary role seems to be discovery) is the ambiguity around the discriminator.  Location of a node with a matching discriminator does not guarantee you've found the node you're looking for.

But if the code that discovers also establishes PASE, then it can hand back a session to the commissioner and commissioner will have very high confidence that it is talking to the node it intends, because successful establishment of PASE means both nodes have the same shared secret.  Furthermore, we always have to establish PASE, so this doesn't introduce more branching in the discoverer.

However, I find it odd that the SetUpCodePairer would have to do any other commissioning steps.  It seems like instead of the SetUpCodePairer consuming a commissioner, it should provide a delegate interface so that it can signal the commissioner when its done.  This would leave the commissioner to make any further commissioning decisions.

msandstedt commented:
> This PR forces getting network credentials before doing discovery and finding out whether they are even needed.
> 
> That's not acceptable, unfortunately. We have to allow commissioners using SetUpCodePairer to do discovery and set up PASE and only then deal with getting network credentials as needed.

Yes, it will not be possible to realize many necessary product flows unless network credentials can be provided dynamically, as needed.  And in many cases this will be after discovery.

However, sorting that out seems like a heavy lift for our feature cut deadline.  If this is a critical feature (in my view it is), can't we bring this through the necessary gatekeeping to potentially continue development and really solve the problem?

cecille commented:
OK, I updated this to be a bit more specific about what each part of the API does. I don't really love the "PairDevice" nomenclature we have going on, but we're kind of past the point of fixing that - it's deeply embedded into the stack. 

I added an option to establish a pase connection though the setup code parier. That gets hooked up thorough chip tool by not specifiying a network technology. I really don't love that, but I also don't really want to rework chip tool to provide an entirely separate command for establishing a pase connection only. I would argue that chip tool is example code anyway, and that can probably be done after TE8.

msandstedt reviewed (approved):
A concern of mine, shared by others, is that we are missing the ability to dynamically inject network credentials in the middle of the commissioning process.  That is something that is important for many UI-based onboarding flows.

However, this PR doesn't create that problem and is adding a necessary capability to chip-tool.  Hence, approve.

cecille commented:
If you want to get information about the device, you can establish a pase connection, read whatever, fill in the required parameters and then commission. This is the same as before, just that the device isn't hanging out in a half-commissioned state with no failsafe running.

bzbarsky-apple commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -83,14 +83,17 @@ CommissioningParameters PairingCommand::GetCommissioningParameters()
-CHIP_ERROR PairingCommand::PairWithManualCode(NodeId remoteId)
-{
-    return CurrentCommissioner().PairDevice(remoteId, mOnboardingPayload);
+    if (mNetworkType == PairingNetworkType::None)
```
So just to check, what codepath will `chip-tool pairing qrcode node-id code` as used in our CI (because we are doing QRCode-based on-network commissioning) take here?  Seems to me like it will take the EstablishPASEConnection path?  If not, why not?  And if it does, what ends up doing the commissioning eventually?

We can change what we run in CI as needed, of course, but I'm not sure whether what we do now will keep working and if not what we should do instead.

cecille replied:
(don't know if this comment was from before the next commit or after)

I noticed the problem with the CI and added some more pairing commands to specify the pase only connection specifically. I THINK that makes it easier, but I'll have to see once the CI runs through.

The discussion of the chip-tool command line arguments and what we're expecting there is somewhat separate from the discussion of the SDK api because I think we're allowed to update chip tool post TE8 branch if we later decide to make that tool more verbose or add options. As long as I can get CI to pass, I don't have any strong opinions on what the chip-tool supports.

bzbarsky-apple replied:
> (don't know if this comment was from before the next commit or after)

Before.  I agree that the key part here is not breaking CI and it was not here how the initial PR was avoiding that.  The new changes make more sense there.

bzbarsky-apple commented on src/controller/CHIPDeviceController.h:
```diff
@@ -537,6 +538,26 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
+     *   Start establishing a PASE connection with a node for the purposes of commissioning.
+     *   Commissioners that wish to use the auto-commissioning functions should use the
+     *   supplied "PairDevice" functions above to automatically establish a connection then
+     *   perform commissioning. This function is intended to be use by commissioners that
```
```suggestion
     *   perform commissioning. This function is intended to be used by commissioners that
```

bzbarsky-apple commented on src/controller/CHIPDeviceController.h:
```diff
@@ -537,6 +538,26 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
+     *   receives the OnPairingComplete call.
+     *
+     *   PASE connections can only be established with nodes that have their commissioning
+     *   window open. The PASE connection will fail if this window is not open and the
```
```suggestion
     *   window open. The PASE connection will fail if this window is not open and in that case
```

bzbarsky-apple commented on src/controller/CHIPDeviceController.h:
```diff
@@ -537,6 +538,26 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
+     * @param[in] remoteDeviceId        The remote device Id.
+     * @param[in] setUpCode             The setup code for connecting to the device
+     */
+    CHIP_ERROR EstablishPASEConnection(NodeId remoteDeviceId, const char * setUpCode);
```
Please check with @sagar-apple as to which Darwin framework callsites that were depending on the previous behavior need to get switched over to this new function?

bzbarsky-apple replied:
Talked to @sagar-apple.  We (Apple) will handle whatever changes to `CHIPDeviceController.mm` are needed here, so you don't have to worry about it.

sagar-apple replied:
They do :) The PairDevice APIs in the CHIPDeviceController.mm need to start calling Establish PASE everywhere instead of the cpp PairDevice APIs.
